src/librustc_trans/mir/lvalue.rs

```diff
@@ -145,10 +145,11 @@ impl<'a, 'tcx> LvalueRef<'tcx> {
         //   * Packed struct - There is no alignment padding
         //   * Field is sized - pointer is properly aligned already
         if st.offsets[ix] == layout::Size::from_bytes(0) || st.packed ||
-            bcx.ccx.shared().type_is_sized(fty) {
-                return (bcx.struct_gep(
-                        ptr_val, adt::struct_llfields_index(st, ix)), alignment);
-            }
+            bcx.ccx.shared().type_is_sized(fty)
```

Comment:
You don't need to do an alignment fixup if you have a foreign vtable. Actually, you only need to do an alignment fixup if the vtable type is `Trait`.

Reply:
Foreign tails were handled by the `!self.has_extra()` a few lines down. I've added a case to the match just beneath, especially since I'm pretty sure the `!self.has_extra()` branch should never be taken.

Reply:
I think it's detritus from old trans, but with your patch it's taken by `Foo<Foo<Opaque>>`. I'm fine with it because you're trying to forbid that.